### PR TITLE
[DPE-3925] Backport juju secrets impl from VM repo

### PR DESCRIPTION
**Pull request by Batalex**
See https://github.com/canonical/zookeeper-operator/pull/129.

I kept the changes as close as possible from the reference PR, except for type hints using modern syntax

### CODE REVIEW

File: src/core/models.py [hunk 1]
```diff
@@ -20,34 +21,64 @@ class StateBase:
...
+        self,
+        relation: Relation,
+        data_interface: Data,
+        component: Unit | Application,
+        substrate: SUBSTRATES,
+    ):
+        super().__init__(relation, data_interface, component, substrate)
+        # Redundant definition as lint can't resolve that super's relation may be None
+        self.relation = relation
+        self.relation_data = cast(dict[str, str], self.data_interface.as_dict(self.relation.id))
```
Comment (Batalex) [inline src/core/models.py:57]: Address https://github.com/canonical/zookeeper-operator/pull/129#discussion_r1555887269

Comment (Batalex) [reply src/core/models.py]: Removed for consistency

File: src/core/cluster.py [hunk 2]
```diff
@@ -6,28 +6,44 @@
...
     # --- RAW RELATION ---
 
     @property
-    def peer_relation(self) -> Relation | None:
+    def peer_relation(self) -> Relation:
         """The cluster peer relation."""
-        return self.model.get_relation(PEER)
+        if not (peer_relation := self.model.get_relation(PEER)):
+            raise AttributeError(f"No peer relation {PEER} found.")
+        return peer_relation
```
Comment (zmraul) [inline src/core/cluster.py:46]: I think this was changed on the VM one.
```python
    @property
    def peer_relation(self) -> Relation | None:
        """The cluster peer relation."""
        return self.model.get_relation(PEER)
```

File: src/core/cluster.py [hunk 3]
```diff
@@ -334,3 +375,10 @@ def ready(self) -> Status:
             return Status.ALL_UNIFIED
 
         return self.stable
+
+    def has_peer_relation(self) -> bool:
+        """The cluster has a peer relation."""
+        try:
+            return bool(self.peer_relation)
+        except AttributeError:
+            return False
```
Comment (zmraul) [inline src/core/cluster.py:384]: This method can be removed and use directly the new boolean override from `RelationState`

**Review by zmraul — CHANGES_REQUESTED** [review]
LGTM! 

I saw a couple of differences on `cluster.py` and `models.py`. Check the diff between zk vm and here, in case there was last changes added to the vm PR. In theory these two files should be the same, so copy/paste is the way to go :)

File: src/core/cluster.py (hunk 3 above)
Comment (Batalex) [reply src/core/cluster.py]: Good calls, I missed those. Fixed on 34dd6d

**Review by zmraul — APPROVED** [review]

**Review by deusebio — APPROVED** [review]
I would also backport to VM the backported changes with the type hints :joy:

**Pull request merged**